### DISCUSSION TITLE
How to get type of loader data

### DISCUSSION THREAD

**Discussion by BrendanC23** [TanStack/router#1732] [Q&A] [answered]
I have a component like this:

```tsx
function Parent() {
    const route = getRouteApi("/");
    const loaderData = route.useLoaderData();

    return (<Child data={loaderData} />);
}

type ChildProps = {
    data: SomeType; // What type is this?
};

function Child({ data }: ChildProps) {
    return null;
}
```

How can I get the type of `route.useLoaderData()`? I've tried `ReturnType<RouteApi<"/">["useLoaderData"]>`, but this gives `unknown`.

I found [this similar discussion](https://github.com/TanStack/router/discussions/1101), but it doesn't have a solution.

**Comment by chuckjrussell** [comment]
I'm also encountering this issue @BrendanC23 did you ever find a solution to this? Temporarily I've added `const data = Route.useLoaderData() as MyDataType` but its obviously a pretty poor solution to the problem.

**Reply by schiller-manuel** [reply]
can you please show a complete example where you need the type?

**Reply by chuckjrussell** [reply]
@schiller-manuel  Thanks for reaching out! Below is an example set of code. I'm still new to tan stack so open to the idea I might be doing this wrong:

``` 
import { getStory } from "@/lib/storyUtils";
import { createFileRoute, Link } from "@tanstack/react-router";

export const Route = createFileRoute("/story/$storyId")({
  component: Story,
  loader: ({ params }) => getStory(params.storyId),
});

function Story() {
  const data = Route.useLoaderData();

  return (
    <div className="p-2">
      <p>{data}</p> //TS Error: Type '{}' is not assignable to type 'ReactNode'.
    </div>
  );
}

```

The data type coming from `Route.useLoaderData()` is just `{}` so I can't accurately make any reference to that data.

I can, as mentioned, cast the data type, so something like `const data = Route.useLoaderData() as Story` but that kind of defeats the purpose of it being strongly typed.

Assuming I'm using the loader right (which of course could very easily be wrong!) what would be good is if `Route.useLoaderData` was typed based on the infered type of the loader property passed in during `createFileRoute`

**Reply by schiller-manuel** [reply]
> Assuming I'm using the loader right (which of course could very easily be wrong!) what would be good is if Route.useLoaderData was typed based on the infered type of the loader property passed in during createFileRoute

It is usually inferred like that. if it is not, then there is an error somewhere else.
can you please create a minimal complete example?

**Reply by Andrewy-gh** [reply]
Piggybacking on this. I am just getting ```any``` on type of ```Route.useLoaderData()```. 

![Screenshot 2025-01-31 194020](https://github.com/user-attachments/assets/eda2ffaf-7745-45ee-bb90-1eb2e28653eb)

![Screenshot 2025-01-31 194056](https://github.com/user-attachments/assets/e8b19520-1ba5-4886-9497-d9d1d68ea4d9)

Here is a minimal example: 

https://stackblitz.com/edit/vitejs-vite-6arcjqkn?file=src%2Froutes%2Findex.tsx

**Reply by schiller-manuel** [reply]
seems to be a stackblitz problem maybe? locally in VS code I see the correct types:

![Screenshot 2025-02-01 at 09 58 44](https://github.com/user-attachments/assets/d0c3a9f9-eaed-46b6-b6ec-1dfbcfa08994)

**Reply by Andrewy-gh** [reply]
Tried locally too. Working now thanks.

**Reply by BrendanC23** [reply]
@schiller-manuel Can you take another look at this question?

Here's a [StackBlitz](https://stackblitz.com/edit/tanstack-router-zsyqcudw?file=src%2Froutes%2Fabout.tsx) that demonstrates the issue. I'm trying to pass part of the loader data down to a child component, but cannot determine the type of the child's props.


```tsx
export const Route = createFileRoute('/about')({
  loader: () => {
    const data = { data: 1, childData: 2 };

    return data;
  },
  component: AboutComponent,
});

const route = getRouteApi('/about');

function AboutComponent() {
  const loaderData = route.useLoaderData();

  return (
      <AboutChild childData={loaderData.childData} />
  );
}

type RouteDataType = ReturnType<(typeof route)['useLoaderData']>; // any
type ChildDataType = RouteDataType['childData']; // This should be number, but is any

type ChildProps = {
  childData: ChildDataType; // How do I determine this type?
};

function AboutChild({ childData }: ChildProps) {
  return (
    <h4>{childData}</h4>
  );
}
```

**Reply by schiller-manuel** [reply, marked as answer]
you can try to get the type of the loader data for a route like this:

```tsx
type LoaderData = RouteById<RegisteredRouter['routeTree'], '/about'>['types']['loaderData']
```

just be aware this uses internal types that might change.
